tests/server/test_workon.py

```diff
@@ -1193,3 +1294,160 @@ async def test_workon_sets_pip_config(
+
+            :param data: The dictionary that will be validated by Pydantic
+            """
+
+            if isinstance(data, dict):
+                optional_fields = [
+                    "started_timestamp",
+                    "completed_timestamp",
+                    "queue_time",
+                    "queue_timeout",
+                    "completion_time",
+                    "completion_timeout",
+                    "exit_code",
+                    "substitute_compile_id",
+                ]
+                for field in optional_fields:
+                    if data[field] == "None":
+                        data[field] = None
+            return data
+
+        def model_post_init(self, __context: Any) -> None:
+            """
+            Add post-validation checks to make sure the provided data is consistent
+            """
+
+            if self.started_timestamp is not None:
+                # This should be computable if requested and started timestamps are defined
+                assert self.queue_time > 0
+                assert self.queue_timeout is None
+            else:
+                assert self.queue_time is None, "Time in queue should not be defined for unstarted compile!"
+                assert self.queue_timeout > 0
+
+            if self.completed_timestamp is not None:
+                # This should be computable if requested and started timestamps are defined
+                assert self.started_timestamp is not None, "Started time should be defined for finished compile!"
+                assert self.completion_time is not None, "Completion time should be defined for finished compile!"
+                assert self.completed_timestamp > self.started_timestamp
+                assert self.completion_timeout is None, "Timed out completion time should not be defined for finished compile!"
+                # assert self.exit_code is not None, "Exit code should be defined for finished compile!"
+            else:
+                assert self.completion_time is None, "Completion time should not be defined for unfinished compile!"
+                assert self.completion_timeout > 0
+                assert self.exit_code is None, "Exit code should not be defined for unfinished compile!"
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        nb_environments: int = 3
+        environments: abc.Sequence[data.model.Environment] = [
+            (
+                await EnvironmentFactory(os.path.join(tmpdirname, f"env-{i}"), project_name=f"project-{i}").create_environment(
+                    name=f"env-{i}"
+                )
+            ).to_dto()
+            for i in range(nb_environments)
+        ]
+
+        for env in environments:
+            result: protocol.Result = await client.notify_change(env.id)
+            assert result.code == 200
+
+        try:
+            raise AssertionError
+        except AssertionError:
```

Comment:
I don't get this part. Is it so we have something written in the logs?

Reply:
Yes, it was written like this to simulate a timeout but I think we could replace it with await asynctio.sleep(..). Wdyt?